telegram/client_e2e_test.go

```diff
@@ -39,7 +39,7 @@ type clientSetup struct {
 }
 
 func testQuorum(
-	trp telegram.Transport,
+	trp *transport.Protocol,
```

Comment:
Why interface got changed to specific type here?

Reply:
Not sure we even need `transport.Protocol` as interface since it encapsulates transport-specific process of connection handshake.

Reply:
Interfaces needed to decouple packages and allow mocking.
See `dcs.Transport`:

```go
// Transport is MTProto connection creator.
type Transport interface {
	Codec() transport.Codec
	Handshake(conn net.Conn) (transport.Conn, error)
}
```

Also note that every method returns interfaces as parameters or return values: `transport.Codec` and `transport.Conn`.

Reply:
Perhaps you are right, but what makes `Transport`/`Protocol` different is that it’s tightly coupled with the socket transport protocol. E.g. it doesn’t make sense to use custom handshake with TCP transport.

Reply:
Also, note that before `dcs` package we didn’t have the `dcs.Transport` *interface*. The transport handled both the protocol handshake (via codec) and dialing network addresses (via underlying transport, e.g. TCP).

Reply:
I've merged your PR and just used `dcs.Protocol` interface to limit scope of `transport.Protocol` (we don't need the `Pipe` method at all).